File: MMORPGServer/Proto/CreateProto/TestProtoProto.cs
//===================================================
//作    者：崔炜斌
//创建时间：2019-12-13 15:41:10
//备    注：
//===================================================
using System.Collections;
using System.Collections.Generic;
using System;

/// <summary>
/// 测试协议
/// </summary>
public struct TestProtoProto : IProto
{
    public ushort ProtoCode { get { return 17001; } }

    public int Id; //
    public string Name; //
    public float price; //
    public int Type; //

    public byte[] ToArray()
    {
        using (MMO_MemoryStream ms = new MMO_MemoryStream())
        {
            ms.WriteUShort(ProtoCode);
            ms.WriteInt(Id);
            ms.WriteUTF8String(Name);
            ms.WriteFloat(price);
            ms.WriteInt(Type);
            return ms.ToArray();
        }
    }

    public static TestProtoProto GetProto(byte[] buffer)
    {
        TestProtoProto proto = new TestProtoProto();
        using (MMO_MemoryStream ms = new MMO_MemoryStream(buffer))
        {
            proto.Id = ms.ReadInt();
            proto.Name = ms.ReadUTF8String();
            proto.price = ms.ReadFloat();
            proto.Type = ms.ReadInt();
        }
        return proto;
    }
}

File: MMORPGServer/Proto/CreateLuaProto/TestProtoProto.lua
--测试协议
TestProtoProto = { ProtoCode = 17001, Id = 0, Name = "", price = 0, Type = 0 }

--这句是重定义元表的索引，就是说有了这句，这个才是一个类
TestProtoProto.__index = TestProtoProto;

function TestProtoProto.New()
    local self = { }; --初始化self
    setmetatable(self, TestProtoProto); --将self的元表设定为Class
    return self;
end


--发送协议
function TestProtoProto.SendProto(proto)

    local ms = CS.LuaHelper.Instance:CreateMemoryStream();
    ms:WriteUShort(proto.ProtoCode);

    ms:WriteInt(proto.Id);
    ms:WriteUTF8String(proto.Name);
    ms:WriteFloat(proto.price);
    ms:WriteInt(proto.Type);

    CS.LuaHelper.Instance:SendProto(ms:ToArray());
    ms:Dispose();
end


--解析协议
function TestProtoProto.GetProto(buffer)

    local proto = TestProtoProto.New(); --实例化一个协议对象
    local ms = CS.LuaHelper.Instance:CreateMemoryStream(buffer);

    proto.Id = ms:ReadInt();
    proto.Name = ms:ReadUTF8String();
    proto.price = ms:ReadFloat();
    proto.Type = ms:ReadInt();

    ms:Dispose();
    return proto;
end